pkg/commands/docker.go

```diff
@@ -243,17 +249,43 @@ func (c *DockerCommand) RefreshContainersAndServices() error {
 	return nil
 }
 
-func (c *DockerCommand) assignContainersToServices(containers []*Container, services []*Service) {
-L:
+func (c *DockerCommand) assignContainersToServices(containers []*Container, services []*Service) ([]*Container, error) {
+	// Get a list of the Container IDs for this Docker Compose project.
+	composeCommand := c.Config.UserConfig.CommandTemplates.DockerCompose
+	output, err := c.OSCommand.RunCommandWithOutput(fmt.Sprintf("%s ps -q", composeCommand))
+	if err != nil {
+		return nil, err
+	}
+
+	// Create an index for Container IDs to make lookups less expensive.
+	lines := utils.SplitLines(output)
+	serviceContainerIds := make(map[string]bool)
+	for _, line := range lines {
+		serviceContainerIds[line] = true
+	}
+
+	// Index Services by name.
+	serviceMap := make(map[string]*Service)
 	for _, service := range services {
-		for _, container := range containers {
-			if !container.OneOff && container.ServiceName == service.Name {
-				service.Container = container
-				continue L
-			}
+		service.Containers = make([]*Container, 0, 1)
+		serviceMap[service.Name] = service
+	}
+
+	// Walk the list of Containers, categorizing them as either associated with a
+	// Service or as standalone.
+	serviceContainerCount := len(serviceContainerIds)
+	standaloneContainerCount := len(containers) - serviceContainerCount
+	standaloneContainers := make([]*Container, 0, standaloneContainerCount)
+	for _, container := range containers {
+		if !container.OneOff && serviceContainerIds[container.ID] {
+			service := serviceMap[container.ServiceName]
```

Comment:
is it possible here that a container may not have a ServiceName set against it? If so, we should add a check for that so that we don't get a nil pointer exception when we say `service.Containers`

Reply:
fair enough